commons/serialization/src/main/scala/sbt/pickling/json/JSONPickleFormat.scala

```diff
@@ -50,33 +53,123 @@ package json {
-      append(s + "\n")
-      pendingIndent = true
+    override def putField(name: String, pickler: (PBuilder) => Unit): PBuilder =
+      if (!isElidedField(name)) {
+        state match {
+          case x: RawEntryState =>
+            x.wasCollectionOrMap = true
+            // Now we know we're in a map state, so we swap into map state.
+            state = MapEntryState(x.previous, x.picklee, x.hints)
+            buf.put("{")
+          case _: MapEntryState =>
+            // here we just need another ,
+            buf.put(",")
+          case _ => sys.error("Cannot put a field when not in entry state!")
+        }
+        // Here we must append all the stringy things around the field.
+        buf.put('"' + name + "\":")
+        pickler(this)
+        this
+      } else this
+    override def endEntry(): Unit = {
+      state match {
+        case RawEntryState(prev, _, _, true) =>
+          // Here we do nothing because it was a collection.
+          state = prev
+        case RawEntryState(prev, picklee, hints, false) =>
+          // Here we have to actually serialize the thing, as we're not a collection or a map.
+          if (primitives.contains(hints.tag.key))
+            primitives(hints.tag.key)(picklee)
+          else if (primitiveArrays.contains(hints.tag.key)) {
+            primitiveArrays(hints.tag.key)(picklee)
+          } else if (isJValue(hints.tag)) {
+            // TODO - Serialize it.
+            import org.json4s.native.JsonMethods._
+            buf.put(compact(render(picklee.asInstanceOf[JValue])))
+          } else {
+            // Note: It's possible the object is empty, so we just put an empty object here,
+            // as the type we're serializing may not have any contents.
+            // we also serialize the "$type" here if needed.
+            // Ignore isStaticallyElidedType. Always output $type.
+            buf.put("{")
+            appendTagString(picklee, hints)
+            buf.put("}")
+          }
+          state = prev
+        case MapEntryState(prev, picklee, hints) =>
+          // Ignore isStaticallyElidedType and always send the $type down.
+          buf.put(",")
+          appendTagString(picklee, hints)
+          buf.put("}")
+          state = prev
+        case RefEntryState(prev) =>
+          state = prev
+        case _ => sys.error("Unable to endEntry() when not in entry state!")
+      }
     }
-    private val tags = new Stack[FastTypeTag[_]]()
+    private def appendTagString(picklee: Any, hints: Hints): Unit =
+      buf.put("\"" + TYPE_TAG_FIELD + "\":\"" + makeTagString(picklee, hints) + "\"")
+    private def makeTagString(picklee: Any, hints: Hints): String =
+      if (hints.tag.key.contains("anonfun$")) picklee.getClass.getName
+      else hints.tag.key
+
+    // We cover ararys of primitives separately here.
+    // NOTE: these are special cased in the core pickler design (probably for binary encoding efficiency)
```

Comment:
This note is about why that isn't handled by a SPickler instead of here in the format.  I think arrays are format primitives because they can be binaray serialized efficiently, while Array[T] uses a pickler.  This is one of those "who should own them" things